### PR TITLE
feat(miner): consider l1msg in txCount

### PR DESCRIPTION
**Pull request by 0xmountaintop**
## 1. Purpose or design rationale of this PR

...

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Review by icemelon — COMMENTED** [review]
remember to update the validation rules

File: miner/worker.go [hunk 1]
```diff
@@ -978,6 +978,12 @@ loop:
 			break
 		}
 		// If we have collected enough transactions then we're done
+		// Originally we only limit l2txs count, but now strictly limit total txs number. 
+		if !w.chainConfig.Scroll.IsValidL2TxCount(w.current.tcount+1) {
```
Comment (colinlyguo) [inline miner/worker.go]:
```suggestion
		if !w.chainConfig.Scroll.IsValidTxCount(w.current.tcount+1) {
```

Comment (colinlyguo) [reply miner/worker.go]: and we can remove `!w.chainConfig.Scroll.IsValidL2TxCount(l2TxCount)` below, and add it back if necessary.

**Review by colinlyguo — DISMISSED** [review]

File: miner/worker.go [hunk 2]
```diff
@@ -978,12 +978,9 @@ loop:
...
 		// If we have collected enough transactions then we're done
-		l2TxCount := w.current.tcount - w.current.l1TxCount
-		if !tx.IsL1MessageTx() { // If the next tx is not L1MessageTx type then +1.
-			l2TxCount++
-		}
-		if !w.chainConfig.Scroll.IsValidL2TxCount(l2TxCount) {
-			log.Trace("Transaction count limit reached", "have", w.current.tcount-w.current.l1TxCount, "want", w.chainConfig.Scroll.MaxTxPerBlock)
+		// Originally we only limit l2txs count, but now strictly limit total txs number.
+		if !w.chainConfig.Scroll.IsValidTxCount(w.current.tcount + 1) {
+			log.Trace("Transaction count limit reached", "have", w.current.tcount+1, "want", w.chainConfig.Scroll.MaxTxPerBlock)
```
Comment (Thegaram) [inline miner/worker.go]:
```suggestion
			log.Trace("Transaction count limit reached", "have", w.current.tcount, "want", w.chainConfig.Scroll.MaxTxPerBlock)
```

File: miner/worker_test.go [hunk 3]
```diff
@@ -680,15 +744,18 @@ func TestL1MsgCorrectOrder(t *testing.T) {
...
 	engine = clique.New(chainConfig.Clique, db)
 
+	maxTxPerBlock := 4
+	chainConfig.Scroll.MaxTxPerBlock = &maxTxPerBlock
 	chainConfig.Scroll.L1Config = &params.L1Config{
 		NumL1MessagesPerBlock: 10,
 	}
 
-	// Insert 3 l1msgs
+	// Insert 4 l1msgs, with one be skipped.
```
Comment (Thegaram) [inline miner/worker_test.go]: Why change this test? Ideally one test should test one behaviour

**Review by Thegaram — DISMISSED** [review]

File: miner/worker_test.go (hunk 3 above)
Comment (colinlyguo) [reply miner/worker_test.go]: because `4 >= l2 msg + l1 non-skip msg && 5 < l2 msg + l1 non-skip msg + l1 skip msg`
so the 4 txs are acceptable.

Comment (colinlyguo) [reply miner/worker_test.go]: reverted in [9f74baf](https://github.com/scroll-tech/go-ethereum/pull/486/commits/9f74baf2fe0afe6a218055f948628f6c273f5a5a)

**Review by Thegaram — DISMISSED** [review]
LGTM, please bump version

**Review by colinlyguo — APPROVED** [review]

**Review by Thegaram — APPROVED** [review]

**Pull request merged**